### PR TITLE
Improve logging of REST request adding MDC

### LINKED ISSUE #3244: Backend logs almost useless
**Describe the bug**
The rest logs are almost useless there are tons of page of 
```
2021-04-21 23:50:02,216 WARN  org.dspace.app.rest.exception.DSpaceApiExceptionControllerAdvice @ An exception has occurred (status:404)
```
sometimes there are error stack trace but most of time there is no reference to which endpoint have generated the stack trace. In previous versions it was possible to identify/guess at the least the Servlet that was invoked as an INFO entry was prosent somewhere before. It was not ideal because with high load concurrent request was hard to distinguish but better than nothing.

In the current main branch the logs don't contain any information about the invoked endpoint, so the only chance is to check the dspace log together with the apache/web log side by side

**To Reproduce**
Browse a dspace 7 repository and check the dspace.log file

**Expected behavior**
The ideal solution would be to have
-  an anonymous and automatic generated request ID included in each log line to keep them together. Each rest call should at least log one entry to note that the call happened;
- at least in the first entry a reference to the page from where the request was generated so that we can quickly identify which page of our angular application use the endpoint. This would apply only to our angular app but maybe could be a suggestion for all the rest client (include a referrer header or something like that)
- expected 404 response as in the case of the configuration endpoint should be identified and reported at info level.

### PR DESCRIPTION
**Pull request by abollini**
## References
* Fixes #3244

## Description
A custom filter has been introduced to add in the thread local a correlation id provided by the client (if any) and an unique request id to simplify the trouble-shooting of the request. This technique is common and well described in the literature https://logging.apache.org/log4j/2.x/manual/thread-context.html
the X-CORRELATION-ID seems to  be often used for this purpose.
An additional custom header X-REFERRER is used instead than the standard (mispelled) REFERER header. The choice of a custom header is due to the fact that many browsers are starting to apply very restrictive policy about the referer header, see https://developers.google.com/web/updates/2020/07/referrer-policy-new-chrome-default
https://web.dev/referrer-best-practices/

This mean that, to get the best from this logging mechanism, the Angular UI should be improved adding to all the requests these two new custom header
- X-CORRELATION-ID with a random generated uuid or any hard to conflict string generated at the application load and keep stable for all the browsing session
- X-REFERRER to note the page that has generated the REST request

**Current limitation**
As the solution is based on a filter that is set after the security filter any log produced by the security filter such as invalid csrf token, bad credentials, etc are not enriched with the extra details.

**REST Contract**
Once we are in agreement about the solution, specifically the name of the Header I will create a PR also for the REST Contract to note the purpose of this two custom headers

## Instructions for Reviewers
Checkout the PR and look to the log file making some activity on the angular UI, the HAL browser or postman.

## Checklist

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [ ] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [n/a] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Comment by tdonohue** [comment]
@abollini : As #3301 was just merged, this PR will need to be rebased on `main` in order to run the newly updated CI build process.

**Comment by abollini** [comment]
@tdonohue done

**Review by tdonohue — COMMENTED** [review]
@abollini : I tested this today just using the HAL Browser, and I found the behavior a bit odd/inconsistent.  Namely, our logs now seem to include two different formats, which will make them very hard to parse consistently. 

It's most easily shown with an example... here's what I see in `dspace.log` when I login via the HAL Browser & then browse to `/api/auth/status`:
```
2021-06-29 16:10:03,845 INFO    org.dspace.authenticate.PasswordAuthentication @ anonymous::authenticate:attempting password auth of user=dspacedemo+admin@gmail.com
2021-06-29 16:10:04,126 INFO    org.dspace.authenticate.PasswordAuthentication @ dspacedemo+admin@gmail.com::authenticate:type=PasswordAuthentication
2021-06-29 16:10:04,127 INFO    org.dspace.eperson.EPersonServiceImpl @ dspacedemo+admin@gmail.com::update_eperson:eperson_id=335647b6-8a52-4ecb-a8c1-7ebabb199bda
2021-06-29 16:10:04,127 INFO    org.dspace.app.rest.security.EPersonRestAuthenticationProvider @ dspacedemo+admin@gmail.com::login:type=explicit
2021-06-29 16:10:05,316 INFO    org.dspace.eperson.EPersonServiceImpl @ dspacedemo+admin@gmail.com::update_eperson:eperson_id=335647b6-8a52-4ecb-a8c1-7ebabb199bda
2021-06-29 16:10:07,789 INFO  unknown 1e374aa7-be4b-4ae8-9419-82fa5dee3cae org.dspace.app.rest.utils.DSpaceAPIRequestLoggingFilter @ Request originated from http://localhost:8080/server/
2021-06-29 16:10:07,789 INFO  unknown 1e374aa7-be4b-4ae8-9419-82fa5dee3cae org.dspace.app.rest.utils.DSpaceAPIRequestLoggingFilter @ Before request [GET /server/api]
2021-06-29 16:10:07,801 INFO  unknown 1e374aa7-be4b-4ae8-9419-82fa5dee3cae org.dspace.app.rest.utils.DSpaceAPIRequestLoggingFilter @ After request [GET /server/api]
2021-06-29 16:10:25,027 INFO  unknown 9b56732a-97a9-46cd-8c13-2f982e074c5a org.dspace.app.rest.utils.DSpaceAPIRequestLoggingFilter @ Request originated from http://localhost:8080/server/
2021-06-29 16:10:25,027 INFO  unknown 9b56732a-97a9-46cd-8c13-2f982e074c5a org.dspace.app.rest.utils.DSpaceAPIRequestLoggingFilter @ Before request [GET /server/api/authn]
2021-06-29 16:10:25,032 INFO  unknown 9b56732a-97a9-46cd-8c13-2f982e074c5a org.dspace.app.rest.utils.DSpaceAPIRequestLoggingFilter @ After request [GET /server/api/authn]
2021-06-29 16:10:27,327 INFO  unknown 91b3c246-c25d-40ca-9636-b0af7b599997 org.dspace.app.rest.utils.DSpaceAPIRequestLoggingFilter @ Request originated from http://localhost:8080/server/
2021-06-29 16:10:27,327 INFO  unknown 91b3c246-c25d-40ca-9636-b0af7b599997 org.dspace.app.rest.utils.DSpaceAPIRequestLoggingFilter @ Before request [GET /server/api/authn/status]
2021-06-29 16:10:27,378 INFO  unknown 91b3c246-c25d-40ca-9636-b0af7b599997 org.dspace.app.rest.utils.DSpaceAPIRequestLoggingFilter @ After request [GET /server/api/authn/status]
```

A few things to notice:
1. Notice how the login request itself is *NOT* logged. You never see the `POST /api/authn/login`. Instead it starts immediately at the `PasswordAuthentication` class
    * I similarly have found the `POST` to logout doesn't get logged either, at least not from the HAL Browser.
2. Notice how the format of the line changes after the first 5 lines.  It seems logs that do *not* come from the REST API have a different format than those that do.  In other words, existing log statements seem to never include a correlation ID or request ID, resulting in two different formats.  Maybe at the very least, these lines should log `unknown unknown`?
3. I find it odd that each request is split into 3 lines...a "Request originated", "Before request" and "After request".  That seems like a lot of unnecessary detail for most production sites.  Why not simplify this into a single line that includes the `referrer` (from the  "Request originated" line) and the request itself (from the "Before request line)?  
    * It seems like it could be one line, similar to: `Request [GET /server/api/authn/status] originating from http://localhost:8080/server/` 
    * Also, I'm not sure that we need both a Before & After log statement...the Before seems "good enough", unless there's a clear reason why both should be needed.
4.  I'm not sure I understand the usefulness of *both* a `correlation-id` and a `request-id` logged for every request (two long UUIDs back to back also results in very long lines in the log).  It seems like the `request-id` doesn't provide us with any additional information at this time (and it doesn't seem to flow through every logged entry based on a single request, or at least I'm not seeing that behavior, for example with the login request flow).  Should we consider just having one of these IDs?  Or maybe use the `X-CORRELATION-ID` if sent, but if not sent, default to creating a server-side `request-id`?

All that said, it's possible this will make more sense once the frontend PR (https://github.com/DSpace/dspace-angular/pull/1255) is working properly, which would make this easier to test.

**Sidenote:** We might consider updating the HAL Browser to also send a `X-CORRELATION-ID`, as it'd make it easier to test with as well.

**Comment by abollini** [comment]
hi @tdonohue thanks for looking to that.
The first two points are related to the current limitation, as noted in the description the login process happen outside the new filter. We can for sure investigate and try to hack a bit the security filter to make sure that the logging filter is invoked or the mdc is properly initialized. As we said that we want to focus on a good enough approch instead than try to get the perfect solution I have preferred to keep this PR simpler.

Point 3 we can for sure merge the before and referrer log together, personally I prefer to have on separate line but no problem to change. We can also easily mute the after but I found it useful as it states that the request has been fully processed and guarantee that we have no stale/zombie request pending.

The request id is the only id that is guaranteed to be present as it is generated by the server and allow us to track a single request. Try to perform an item creation or a community delete to see more logs. Of course if you are the only one browsing the app it doesn't seem very useful but it is a real service with simultanious users.

The request and correlation id are both required and useful when available ad angular itself is the main source of concurrency as the browser process many request in parallel.

**Comment by tdonohue** [comment]
@abollini : thanks for the quick response! 

I had missed the "current limitation" note above about how this doesn't quite work for login/logout, invalid CSRF, etc.  That's completely reasonable as this just should be a "quick fix" as you noted.  One option we could consider is to change the log4j2.xml configuration slightly to:
```
$${ctx:correlationID:-unknown} $${ctx:requestID:-unknown}
```
This would result in lines _without_ a `correlationID` or `requestID` to automatically have `unknown unknown` in the log line.  I admit though that this results in a large number of log lines with `unknown unknown` when DSpace first boots up.  But, it keeps the log format more consistent.  That said, if you prefer the current format, I don't feel strongly in requiring this change...just presenting it as an option.

In terms of the number of lines per request, I still prefer myself a shorter syntax (as noted in 3 above).  However, we could make this easily configurable via something like `logging.server.include-after-request`.  When set to `true`, you see that `After request` line in the logs.  When set to `false`, you just see a **single line** which says something like `Request [GET /server/api/authn/status] originating from http://localhost:8080/server/`.    I'd recommend `false` by default, but if others disagree, we could even default to `true`.

Regarding 4, I see your point now after playing with this more.  The `correlationID` and `requestID` do serve different purposes.  I'm OK with leaving both in for now...we can always improve more later.

(As a sidenote, in last week's meeting, @artlowel noted we may want to look closely at whether the `correlationID` aligns with GDPR / privacy restrictions.  If it is easy to map it back to a *specific user/login* (which it likely is as soon as you login), then it may not be an anonymous identifier.  This doesn't mean we need to change this PR, but it does mean we may need to determine whether we need a way to still "anonymize" these logs via a command-line script or similar... I'd leave that out of scope for 7.0 though, but something we may need to think about for 7.1.)

All in all, the direction of this PR looks good, as does the code.  The only reason I haven't given it a thumbs up, is that I'd like to ensure it is tested with the corresponding Angular PR (and that other PR is not working at this time).

**Comment by LucaGiamminonni** [comment]
After the suggestions I added a commit for:
- change the log layout to automatically have unknown unknown
- add the two headers to the allowed headers in the CORS configuration
- merge the two before request log lines into one

File: dspace/config/log4j2.xml
Comment (tdonohue) [inline dspace/config/log4j2.xml:32]: Unfortunately, I forgot that this `log4j2.xml` file is [filtered by Ant](https://github.com/DSpace/DSpace/blob/main/dspace/src/main/config/build.xml#L355), which means we MUST escape these dollar signs. I tried this out locally, and below is the correct syntax, as Ant transforms a `$$` into a literal `$`.
```
<!-- NOTE: this pattern uses four dollar signs to escape property substitution by the Ant filter
     process used to install this file. Once installed, only two dollar signs will exist. -->
<Layout type='PatternLayout'
        pattern='%d %-5p $$$${ctx:correlationID:-unknown} $$$${ctx:requestID:-unknown} %c @ %m%n'/>
```

**Review by tdonohue — CHANGES_REQUESTED** [review]
@LucaGiamminonni and @abollini :  First off, this is **almost ready**.  However, two quick comments:

1. I found a bug in that we need to escape dollar signs in the `log4j2.xml` file, as it is filtered by Ant (see inline comment below for how to correct this)
2. _(Not required)_ I still feel we may want to make the `After request` log line *disabled by default*.  In testing this with the UI, I'm finding _some_ UI pages trigger 10-15 REST Calls... with 2 lines per REST call, that's (at least) 20-30 log lines for a *single UI page access*.  If you had just 10 users accessing your site simultaneously, **you may generate 200-300 log lines in a few seconds.**  That's a lot of logs, and we could cut it in half by disabling the "After request" line.
    * As noted in my previous comment, we could easily make this configurable.  Create a new configuration `logging.server.include-after-request = false`, and check its value in the `afterRequest()` method to determine whether to include that INFO line. If false, don't log it. If true, log it.
    
Overall, I'm +1 this as long as the bug below is fixed.  If you disagree strongly about making the `After request` log line configurable, we can leave it as is.  I just wanted to make sure you realized how many lines of logs are generated just from one page in the UI.

File: dspace/config/log4j2.xml
Comment (LucaGiamminonni) [reply dspace/config/log4j2.xml]: Hi @tdonohue, thanks for pointing out the missing escape. I fixed the bug and also added the logging.server.include-after-request property to enable/disable log after request.

**Comment by LucaGiamminonni** [comment]
Hi @tdonohue, thanks for pointing out the missing escape. I fixed the bug and also added the logging.server.include-after-request property to enable/disable log after request.

**Review by tdonohue — APPROVED** [review]
👍 Looks good to me now. Thanks @LucaGiamminonni and @abollini !

@benbosman : Did you or someone from Atmire want to give this a quick look, or should I just merge this with one approval?

**Comment by tdonohue** [comment]
Merging as-is, since @artlowel [approved the Angular side of things](https://github.com/DSpace/dspace-angular/pull/1255#pullrequestreview-710619003)... This was just waiting on a general approval of the direction by someone else.

Thanks again @LucaGiamminonni and @abollini

**Pull request merged**
